server/core/region.go

```diff
@@ -608,7 +608,45 @@ func (r *RegionsInfo) AddRegion(region *RegionInfo) []*RegionInfo {
 	}
 
 	r.regions.Put(region)
+	r.addRegionToSubTree(region)
+	return overlaps
+}
+
+// ReplaceOrAddRegion adds or replaces RegionInfo to regionTree and regionMap, also update leaders and followers by region peers
+func (r *RegionsInfo) ReplaceOrAddRegion(region *RegionInfo, overlaps []*RegionInfo, oldRegionItem *regionItem) {
+	// If oldRegionItem is nil, or ID of the region not equal ID of the oldRegionItem's region.
+	if !isEqualRegion(region, oldRegionItem) {
+		// Remove all the regions of the overlaps.
+		for _, item := range overlaps {
+			r.RemoveRegion(r.GetRegion(item.GetID()))
+		}
+		// If old region exists. remove too.
+		if origin := r.regions.Get(region.GetID()); origin != nil {
+			r.RemoveRegion(origin)
+		}
+		// Add the new region to the regionTree.
+		r.tree.update(region)
+	} else {
+		// When else, remove the regions of the overlaps except the old region.
```

Comment:
Can you add a test about this logic?
```
There is a situation:
Region:   |____________2______________|
Region:                 |______2______|(new)
````
the old region 2 split to two reigons , and new region 2(has same ID) coming from heartbeat.
In the situation, will remove the old region cause by `overlaps` contains the old region, but the update(line 639) is in the `oldRegionItem`(that already remove from the tree).

BTW, Does the main optimize is the `removeRegionFromSubTree`?